core/src/main/scala/kafka/log/Log.scala

```diff
@@ -470,22 +470,44 @@ class Log(val dir: File,
   def read(startOffset: Long, maxLength: Int, maxOffset: Option[Long] = None): FetchDataInfo = {
     trace("Reading %d bytes from offset %d in log %s of length %d bytes".format(maxLength, startOffset, name, size))
 
-    // check if the offset is valid and in range
-    val next = nextOffsetMetadata.messageOffset
-    if(startOffset == next)
-      return FetchDataInfo(nextOffsetMetadata, MessageSet.Empty)
-    
+    // We first create a few local variables to avoid race conditions with updates to the log. The order matters here.
+    // We can allow currentLastEntry to have a newer value than currentNextOffsetMetadata but not vice versa.
+    val currentNextOffsetMetadata = nextOffsetMetadata
+    val currentNext = currentNextOffsetMetadata.messageOffset
+    val currentLastEntry = segments.lastEntry
+
+    if(startOffset == currentNext)
+      return FetchDataInfo(currentNextOffsetMetadata, MessageSet.Empty)
+
     var entry = segments.floorEntry(startOffset)
-      
+
     // attempt to read beyond the log end offset is an error
-    if(startOffset > next || entry == null)
-      throw new OffsetOutOfRangeException("Request for offset %d but we only have log segments in the range %d to %d.".format(startOffset, segments.firstKey, next))
+    if(startOffset > currentNext || entry == null)
+      throw new OffsetOutOfRangeException(("Request for offset %d but we only have log segments in the range %d " +
+        "to %d.").format(startOffset, segments.firstKey, currentNext))
     
-    // do the read on the segment with a base offset less than the target offset
+    // Do the read on the segment with a base offset less than the target offset
     // but if that segment doesn't contain any messages with an offset greater than that
     // continue to read from successive segments until we get some messages or we reach the end of the log
+    // Note that at this point startOffset < currentNext. There are three cases:
```

Comment:
It's useful to describe the handling of those corner cases. But it's more important to explain why we need to have maxPosition in the first place.